lib/matplotlib/widgets.py

```diff
@@ -2003,117 +2051,254 @@ def new_axes(self, ax):
+        # Note: event.xdata/ydata in data coordinates, event.x/y in pixels
+        e_idx, e_dist = self._edge_handles.closest(event.x, event.y)
 
-        self.prev = x, y
+        # Prioritise center handle over other handles
+        # Use 'C' to match the notation used in the RectangleSelector
+        if 'move' in self.state:
+            self._active_handle = 'C'
+        elif e_dist > self.maxdist:
+            # Not close to any handles
+            self._active_handle = None
+            if self.drag_from_anywhere and self._contains(event):
+                # Check if we've clicked inside the region
+                self._active_handle = 'C'
+                self._extents_on_press = self.extents
+            else:
+                self._active_handle = None
+                return
+        else:
+            # Closest to an edge handle
+            self._active_handle = self._edge_order[e_idx]
+
+        # Save coordinates of rectangle at the start of handle movement.
+        self._extents_on_press = self.extents
+
+    def _contains(self, event):
+        """Return True if event is within the patch."""
+        return self._rect.contains(event, radius=0)[0]
+
+    @property
+    def extents(self):
+        """Return extents of the span selector."""
         if self.direction == 'horizontal':
-            v = x
+            vmin = self._rect.get_x()
+            vmax = vmin + self._rect.get_width()
         else:
-            v = y
+            vmin = self._rect.get_y()
+            vmax = vmin + self._rect.get_height()
+        return vmin, vmax
+
+    @extents.setter
+    def extents(self, extents):
+        # Update displayed shape
+        self._draw_shape(*extents)
+        # Update displayed handles
+        self._edge_handles.set_data(self.extents)
+        self.set_visible(self.visible)
+        self.update()
+
+
+class ToolLineHandles:
+    """
+    Control handles for canvas tools.
+
+    Parameters
+    ----------
+    ax : `matplotlib.axes.Axes`
+        Matplotlib axes where tool handles are displayed.
+    positions : 1D array
+        Positions of handles in data coordinates.
+    direction : {"horizontal", "vertical"}
+        Direction of handles, either 'vertical' or 'horizontal'
+    line_props : dict
+        Additional line properties. See `matplotlib.lines.Line2D`.
+    """
 
-        minv, maxv = v, self.pressv
-        if minv > maxv:
-            minv, maxv = maxv, minv
+    def __init__(self, ax, positions, direction, line_props=None,
+                 useblit=True):
+        self.ax = ax
+
+        _api.check_in_list(['horizontal', 'vertical'], direction=direction)
+        self._direction = direction
+
+        if line_props is None:
+            line_props = {}
+        line_props.update({'visible': False, 'animated': useblit})
+
+        line_fun = ax.axvline if self.direction == 'horizontal' else ax.axhline
+
+        self.artists = [line_fun(p, **line_props) for p in positions]
```

Comment:
Normally, it is one or another, depending if the span selector is vertical or horizontal, so we need to loop over only one of the two.

Reply:
I meant, can one of those functions replace the loop? It might not be the case, since the return type is different.

Reply:
OK, you meant something along these lines:
```python
import matplotlib.pyplot as plt

fig = plt.figure()
ax = fig.add_subplot()
ax.plot(range(100))

positions = (10, 20)

# instead of having a loop over the positions, pass the positions tuple as argument
ax.axvline(x=positions)
```

However, the first argument of `ax.axvline` or `ax.axhline` expect a float and we can't pass an iterable. The alternative would be to use `ax.vlines` or `ax.hlines` with a transform for the infinite direction, but it will more complicated.